pkg/pgmodel/series_set.go

```diff
@@ -0,0 +1,196 @@
+	p.err = errInvalidData
+
+	if len(arrays) != 4 {
+		return nil
+	}
+
+	ps := &pgxSeries{}
+
+	if err := ps.labelNames.DecodeBinary(nil, arrays[0]); err != nil {
+		return nil
+	}
+	if err := ps.labelValues.DecodeBinary(nil, arrays[1]); err != nil {
+		return nil
+	}
+
+	if len(ps.labelNames.Elements) != len(ps.labelValues.Elements) {
+		return nil
+	}
+
+	if err := ps.times.DecodeBinary(nil, arrays[2]); err != nil {
+		return nil
+	}
+
+	if err := ps.values.DecodeBinary(nil, arrays[3]); err != nil {
+		return nil
+	}
+
+	if len(ps.times.Elements) != len(ps.values.Elements) {
+		return nil
+	}
+
+	p.err = nil
+	return ps
+}
+
+// Err implements storage.SeriesSet.
+func (p *pgxSeriesSet) Err() error {
+	return p.err
+}
+
+// pgxSeries implements storage.Series.
+type pgxSeries struct {
+	labelNames  pgtype.TextArray
+	labelValues pgtype.TextArray
+	times       pgtype.TimestamptzArray
+	values      pgtype.Float8Array
+}
+
+// Labels returns the label names and values for the series.
+func (p *pgxSeries) Labels() labels.Labels {
+	ll := make(labels.Labels, len(p.labelNames.Elements))
+
+	for i := range ll {
+		ll[i].Name = p.labelNames.Elements[i].String
+		ll[i].Value = p.labelValues.Elements[i].String
+	}
+
+	sort.Sort(ll)
+
+	return ll
+}
+
+// Iterator returns a chunkenc.Iterator for iterating over series data.
+func (p *pgxSeries) Iterator() chunkenc.Iterator {
+	return newIterator(p.times, p.values)
+}
+
+// pgxSeriesIterator implements storage.SeriesIterator.
+type pgxSeriesIterator struct {
+	cur          int32
+	totalSamples int32
+	times        pgtype.TimestamptzArray
+	values       pgtype.Float8Array
+}
+
+// newIterator returns an iterator over the samples. It expects times and values to be the same length.
+func newIterator(times pgtype.TimestamptzArray, values pgtype.Float8Array) *pgxSeriesIterator {
+	return &pgxSeriesIterator{
+		cur:          -1,
+		totalSamples: int32(len(times.Elements)),
+		times:        times,
+		values:       values,
+	}
+}
+
+// Seek implements storage.SeriesIterator.
+func (p *pgxSeriesIterator) Seek(t int64) bool {
+	found := false
+	p.cur = 0
+
+	for !found && p.cur < p.totalSamples {
+		if p.getTs() >= t {
+			found = true
+			break
+		}
+		p.cur++
+	}
+
+	return found
+}
+
+func (p *pgxSeriesIterator) getTs() int64 {
+	v := p.times.Elements[p.cur]
+
+	if v.Status == pgtype.Null {
+		return 0
+	}
+
+	switch v.InfinityModifier {
+	case pgtype.NegativeInfinity:
+		return math.MinInt64
+	case pgtype.Infinity:
+		return math.MaxInt64
+	default:
+		return v.Time.UnixNano() / 1e6
+	}
+}
+
+func (p *pgxSeriesIterator) getVal() float64 {
+	return p.values.Elements[p.cur].Float
+}
+
+// At implements storage.SeriesIterator.
+func (p *pgxSeriesIterator) At() (t int64, v float64) {
+	if p.cur >= p.totalSamples {
+		return 0, 0
+	}
+	return p.getTs(), p.getVal()
+}
+
+// Next implements storage.SeriesIterator.
+func (p *pgxSeriesIterator) Next() bool {
```

Comment:
The right thing to do here is probably iterate until both the Time and  Value is not null.